tests/unit/lib/secrets/vault_csi_tests.py

```diff
@@ -0,0 +1,205 @@
+    write_secrets(data_path, data)
+    write_symlinks(data_path)
+    return csi_dir
+
+
+def simulate_secret_update(
+    csi_dir: Path, updated_data: typing.Optional[typing.Dict[str, SecretType]] = None
+) -> None:
+    """Simulates either TTL expiry / a secret update."""
+    old_data_path = csi_dir.joinpath("..data").resolve()
+    # Clone the data directory
+    new_data_path = Path(csi_dir, f'..{datetime.datetime.today().strftime("%Y_%m_%d_%H_%M_%S.%f")}')
+    # Update the secret
+    if updated_data:
+        write_secrets(new_data_path, updated_data)
+    else:
+        shutil.copytree(old_data_path, new_data_path)
+    write_symlinks(new_data_path)
+    shutil.rmtree(old_data_path)
+
+
+def get_secrets_store(csi_dir: str) -> SecretsStore:
+    store = secrets_store_from_config({"secrets.path": csi_dir, "secrets.provider": "vault_csi"})
+    assert isinstance(store, VaultCSISecretsStore)
+    return store
+
+
+EXAMPLE_SECRETS_DATA = {
+    "secret/example-service/example-secret": {
+        "request_id": "8487d906-2154-0151-d07e-57f41447326a",
+        "lease_id": "",
+        "lease_duration": 2764800,
+        "renewable": False,
+        "data": {"password": "password", "type": "credential", "username": "reddit"},
+        "warnings": None,
+    },
+    "secret/example-service/nested/example-nested-secret": {
+        "request_id": "8487d906-2154-0151-d07e-57f41447326a",
+        "lease_id": "",
+        "lease_duration": 2764800,
+        "renewable": False,
+        "data": {"password": "password", "type": "credential", "username": "reddit"},
+        "warnings": None,
+    },
+    "secret/bare-secret": {
+        "request_id": "8487d906-2154-0151-d07e-57f41447326a",
+        "lease_id": "",
+        "lease_duration": 2764800,
+        "renewable": False,
+        "data": {"password": "password", "type": "credential", "username": "reddit"},
+        "warnings": None,
+    },
+}
+
+EXAMPLE_UPDATED_SECRETS = EXAMPLE_SECRETS_DATA.copy()
+EXAMPLE_UPDATED_SECRETS.update(
+    {
+        "secret/example-service/example-secret": {
+            "request_id": "8487d906-2154-0151-d07e-57f41447326a",
+            "lease_id": "",
+            "lease_duration": 2764800,
+            "renewable": False,
+            "data": {
+                "password": "new_password",
+                "type": "credential",
+                "username": "new_reddit",
+            },
+            "warnings": None,
+        },
+    }
+)
+
+
+class StoreTests(unittest.TestCase):
+    def setUp(self):
+        self.csi_dir = new_fake_csi(EXAMPLE_SECRETS_DATA)
+
+    def tearDown(self):
+        shutil.rmtree(self.csi_dir)
+
+    def test_can_load_credential_secret(self):
+        secrets_store = get_secrets_store(str(self.csi_dir))
+        data = secrets_store.get_credentials("secret/example-service/example-secret")
+        assert data.username == "reddit"
+        assert data.password == "password"
```

Comment:
can you test all of the secret types?

(does Python not automatically handle un-base64-ing secrets that have `encoding: base64`?)

Reply:
All of them are now tested. Looks to me like only 2 types support base64 decoding: Simple and versioned.

Reply:
Yep, that's correct.